### DISCUSSION TITLE
How does aioquic handle IP addr changes?

### DISCUSSION THREAD

**Discussion by hari-19** [aiortc/aioquic#491] [Q&A] [answered]
QUIC is supposed to have connection migration benefits. How is it implemented in aioquic. Are there any APIs to manually trigger and IP change in a multi-IP device?

**Comment by rthalley** [comment, marked as answer]
aioquic follows the protocol (RFC 9000 section 9).  Basically if a client changes the address it is sending from, the connection will migrate automatically.  In the asyncio implementation that aioquic provides, there is no API for changing the socket, so you can't trigger manually.   If you took more control of the I/O yourself and just used aioquic's I/O-less state machine, then if you changed the address you were sending on it should migrate automatically, but I haven't tried this myself.

**Reply by hari-19** [reply]
Yes. While working with the aioquic's source code, I noticed that if the client changes its address, the connection migrates automatically. 

I was more interested in how a network (IP) change at the client's end device can be handled gracefully. The current implementation seems to bind a single UDP socket to the client. It would be nice to have something like an automatic socket change if the network address (IP) changes, as this is one of the major advantages of QUIC's connection migration.

Thanks for the prompt response. I have tried changing the socket on the client's end between the connection and it seems to work.

**Reply by rthalley** [reply]
Looking more at the code, I see that it makes its own socket, and this socket is bound to a wild local address and is not connected.  So, I think this means that you get automatic migration if the system's IP address changes, or changes in routing cause the packet to go out via a different interface.

**Reply by hari-19** [reply]
I overlooked the wild local address binding. Thanks for pointing it out !!
